From be2c3d0d315fe944368c5a5b83bb7f59fcb2a4ea Mon Sep 17 00:00:00 2001
From: Ken Sharp <ken.sharp@artifex.com>
Date: Mon, 21 Nov 2022 13:46:15 +0000
Subject: [PATCH] Allow @file syntax to include files in the ROM file system

This started as a customer request, where the customer wanted to
store a file in the ROM file system with a bunch of command line
parameters in it, and then use the @file syntax to run that file
and setup the interpreter. In essence a 'canned' setup stored in
the ROM file system.

That didn't work, at all, because of a dichotomy in the logic of
argument processing. The code used 'lib_file_open' in order to find
the file using the search path which, on Ghostscript, includes all
the iodevices, Rom file system, RAM file system etc.

That function returns a stream which the calling function 'lib_fopen'
would then reach inside and pull out the pointer to the underlying
gp__file *, and return, because the argument processing was written
to use a gp_file *, not a stream *.

But.... For the ROM file system, what we store in the 'file' member
of the stream is not a gp_file *, its a 'node' *. Trying to use that
as a gp_file * rapidly leads to a seg fault.

So this commit reworks the argument processing to use a stream *
instead of a gp_file *. In reality there is no real difference to the
logic we simply use a different call. There are a lot of fiddly
references to change unfortunately.

There are some consequences; we need to pass stream * around instead of
gp_file *, and in particular the 'encoding' functions, which are
OS-specific need to use a stream *. This isn't a problem for the
interpreters and graphics library, but mkromfs *also* uses those
OS-specific files, and it does not have the stream library available.

Rather than trying to include it, we note that we don't actually need
to use these functions for mkromfs and we do the same hackery as for
other missing functionality; we define a stub function that does nothing
but will permit the compile to complete.
---
 base/gp.h         |  4 ++--
 base/gp_unix.c    | 13 +++++++------
 base/gp_win32.c   | 13 +++++++------
 base/gsargs.c     | 26 +++++++++++++-------------
 base/gsargs.h     | 13 +++++++------
 base/lib.mak      |  1 +
 base/mkromfs.c    | 13 +++++++++++++
 base/unix-aux.mak |  4 ++--
 base/winplat.mak  |  6 +++---
 pcl/pl/pl.mak     |  4 ++--
 pcl/pl/plapi.c    | 17 +++++++++--------
 pcl/pl/plmain.c   |  8 ++++----
 pcl/pl/plmain.h   |  2 +-
 psi/files.h       |  4 ++--
 psi/imainarg.c    |  8 ++++----
 psi/imainarg.h    |  4 ++--
 psi/iminst.h      |  4 ++--
 psi/psapi.c       | 16 ++++++++--------
 psi/zfile.c       | 18 +++---------------
 19 files changed, 92 insertions(+), 86 deletions(-)

diff --git a/base/gp.h b/base/gp.h
index 33a35cccc..993702f6b 100644
--- a/base/gp.h
+++ b/base/gp.h
@@ -690,7 +690,7 @@ void gp_enumerate_files_close_impl(gs_memory_t *memory, file_enum * pfen);
  * a unicode codepoint. Returns EOF for end of file (or string).
  */
 int
-gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr);
+gp_local_arg_encoding_get_codepoint(stream *s, const char **astr);
 
 int
 gp_xpsprint(char *filename, char *printername, int *result);
diff --git a/base/gp_unix.c b/base/gp_unix.c
index 98db2ddf8..73271e770 100644
--- a/base/gp_unix.c
+++ b/base/gp_unix.c
@@ -25,6 +25,7 @@
 #include "gx.h"
 #include "gsexit.h"
 #include "gp.h"
+#include "stream.h"
 
 #ifdef HAVE_FONTCONFIG
 #  include <fontconfig/fontconfig.h>
@@ -497,7 +498,7 @@ void gp_enumerate_fonts_free(void *enum_state)
 
 #ifdef __MINGW32__
 int
-gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
+gp_local_arg_encoding_get_codepoint(stream *s, const char **astr)
 {
     int len;
     int c;
@@ -505,8 +506,8 @@ gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
     wchar_t unicode[2];
     char utf8[4];
 
-    if (file) {
-        c = fgetc(file);
+    if (s) {
+        c = spgetc(s);
         if (c == EOF)
             return EOF;
     } else if (**astr) {
@@ -519,8 +520,8 @@ gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
 
     arg[0] = c;
     if (IsDBCSLeadByte(c)) {
-        if (file) {
-            c = fgetc(file);
+        if (s) {
+            c = spgetc(s);
             if (c == EOF)
                 return EOF;
         } else if (**astr) {
diff --git a/base/gp_win32.c b/base/gp_win32.c
index e04a7e457..1142a8250 100644
--- a/base/gp_win32.c
+++ b/base/gp_win32.c
@@ -24,6 +24,7 @@
 #include "gsmemory.h"		/* for gp.h */
 #include "gserrors.h"
 #include "gp.h"
+#include "stream.h"
 
 /* ------ Miscellaneous ------ */
 
@@ -129,7 +130,7 @@ const char gp_current_directory_name[] = ".";
 
 #if defined(__WIN32__) && !defined(METRO)
 int
-gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
+gp_local_arg_encoding_get_codepoint(stream *s, const char **astr)
 {
     int len;
     int c;
@@ -137,8 +138,8 @@ gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
     wchar_t unicode[2];
     char utf8[4];
 
-    if (file) {
-        c = gp_fgetc(file);
+    if (s) {
+        c = spgetc(s);
         if (c == EOF)
             return EOF;
     } else if (**astr) {
@@ -151,8 +152,8 @@ gp_local_arg_encoding_get_codepoint(gp_file *file, const char **astr)
 
     arg[0] = c;
     if (IsDBCSLeadByte(c)) {
-        if (file) {
-            c = gp_fgetc(file);
+        if (s) {
+            c = spgetc(s);
             if (c == EOF)
                 return EOF;
         } else if (**astr) {
diff --git a/base/gsargs.c b/base/gsargs.c
index a17998f95..d2402563d 100644
--- a/base/gsargs.c
+++ b/base/gsargs.c
@@ -59,7 +59,7 @@ int codepoint_to_utf8(char *cstr, int rune)
     return idx;
 }
 
-static int get_codepoint_utf8(gp_file *file, const char **astr)
+static int get_codepoint_utf8(stream *s, const char **astr)
 {
     int c;
     int rune;
@@ -72,7 +72,7 @@ static int get_codepoint_utf8(gp_file *file, const char **astr)
      * what they get. */
 
     do {
-        c = (file ? gp_fgetc(file) : (**astr ? (int)(unsigned char)*(*astr)++ : EOF));
+        c = (s ? spgetc(s) : (**astr ? (int)(unsigned char)*(*astr)++ : EOF));
         if (c == EOF)
             return EOF;
         if (c < 0x80)
@@ -93,7 +93,7 @@ lead: /* We've just read a byte >= 0x80, presumably a leading byte */
         else
             continue; /* Illegal - skip it */
         do {
-            c = (file ? gp_fgetc(file) : (**astr ? (int)(unsigned char)*(*astr)++ : EOF));
+            c = (s ? spgetc(s) : (**astr ? (int)(unsigned char)*(*astr)++ : EOF));
             if (c == EOF)
                 return EOF;
             rune = (rune<<6) | (c & 0x3f);
@@ -120,9 +120,9 @@ int
 arg_init(arg_list     * pal,
          const char  **argv,
          int           argc,
-         gp_file      *(*arg_fopen)(const char *fname, void *fopen_data),
+         stream      *(*arg_fopen)(const char *fname, void *fopen_data),
          void         *fopen_data,
-         int           (*get_codepoint)(gp_file *file, const char **astr),
+         int           (*get_codepoint)(stream *s, const char **astr),
          gs_memory_t  *memory)
 {
     int code;
@@ -183,7 +183,7 @@ arg_finit(arg_list * pal)
         arg_source *pas = &pal->sources[pal->depth--];
 
         if (pas->is_file)
-            gp_fclose(pas->u.file);
+            sclose(pas->u.strm);
         else if (pas->u.s.memory)
             gs_free_object(pas->u.s.memory, pas->u.s.chars, "arg_finit");
     }
@@ -191,10 +191,10 @@ arg_finit(arg_list * pal)
 
 static int get_codepoint(arg_list *pal, arg_source *pas)
 {
-    int (*fn)(gp_file *file, const char **str);
+    int (*fn)(stream *s, const char **str);
 
     fn = (!pas->is_file && pas->u.s.decoded ? get_codepoint_utf8 : pal->get_codepoint);
-    return fn(pas->is_file ? pas->u.file : NULL, &pas->u.s.str);
+    return fn(pas->is_file ? pas->u.strm : NULL, &pas->u.s.str);
 }
 
 /* Get the next arg from a list. */
@@ -251,7 +251,7 @@ arg_next(arg_list * pal, const char **argstr, const gs_memory_t *errmem)
             if (c == EOF) {
                 /* EOF before any argument characters. */
                 if (pas->is_file)
-                    gp_fclose(pas->u.file);
+                    sclose(pas->u.strm);
                 else if (pas->u.s.memory)
                     gs_free_object(pas->u.s.memory, pas->u.s.chars,
                                    "arg_next");
@@ -391,7 +391,7 @@ arg_next(arg_list * pal, const char **argstr, const gs_memory_t *errmem)
          * it to the caller. */
         if (pal->expand_ats && **argstr == '@') {
             char *fname;
-            gp_file *f;
+            stream *s;
             if (pal->depth+1 == arg_depth_max) {
                 errprintf(errmem, "Too much nesting of @-files.\n");
                 return_error(gs_error_Fatal);
@@ -401,15 +401,15 @@ arg_next(arg_list * pal, const char **argstr, const gs_memory_t *errmem)
             if (gs_add_control_path(pal->memory, gs_permit_file_reading, fname) < 0)
                 return_error(gs_error_Fatal);
 
-            f = (*pal->arg_fopen) (fname, pal->fopen_data);
+            s = (*pal->arg_fopen) (fname, pal->fopen_data);
             DISCARD(gs_remove_control_path(pal->memory, gs_permit_file_reading, fname));
-            if (f == NULL) {
+            if (s == NULL) {
                 errprintf(errmem, "Unable to open command line file %s\n", *argstr);
                 return_error(gs_error_Fatal);
             }
             pas = &pal->sources[++pal->depth];
             pas->is_file = true;
-            pas->u.file = f;
+            pas->u.strm = s;
             *argstr = NULL; /* Empty the argument string so we don't return it. */
             continue; /* Loop back to parse the first arg from the file. */
         }
diff --git a/base/gsargs.h b/base/gsargs.h
index 4767c85ae..36b2055f8 100644
--- a/base/gsargs.h
+++ b/base/gsargs.h
@@ -20,6 +20,7 @@
 #  define gsargs_INCLUDED
 
 #include "std.h"
+#include "stream.h"
 
 /*
  * We need to handle recursion into @-files.
@@ -39,14 +40,14 @@ typedef struct arg_source_s {
             gs_memory_t *memory;/* if non-0, free chars when done with it */
             const char *str;    /* string being read */
         } s;
-        gp_file *file;
+        stream *strm;
     } u;
 } arg_source;
 typedef struct arg_list_s {
     bool expand_ats;            /* if true, expand @-files */
-    gp_file *(*arg_fopen) (const char *fname, void *fopen_data);
+    stream *(*arg_fopen) (const char *fname, void *fopen_data);
     void *fopen_data;
-    int (*get_codepoint)(gp_file *file, const char **astr);
+    int (*get_codepoint)(stream *s, const char **astr);
     gs_memory_t *memory;
     const char **argp;
     int argn;
@@ -61,9 +62,9 @@ int codepoint_to_utf8(char *cstr, int rune);
 int arg_init(arg_list    *pal,
              const char **argv,
              int          argc,
-             gp_file     *(*arg_fopen)(const char *fname, void *fopen_data),
+             stream      *(*arg_fopen)(const char *fname, void *fopen_data),
              void        *fopen_data,
-             int          (*get_codepoint)(gp_file *file, const char **astr),
+             int          (*get_codepoint)(stream *s, const char **astr),
              gs_memory_t *mem);
 
 /*
diff --git a/base/lib.mak b/base/lib.mak
index 682457829..a927afc12 100644
--- a/base/lib.mak
+++ b/base/lib.mak
@@ -4078,6 +4078,7 @@ $(GLSRC)gsalloc.h:$(GLGEN)arch.h
 $(GLSRC)gsargs.h:$(GLSRC)std.h
 $(GLSRC)gsargs.h:$(GLSRC)stdpre.h
 $(GLSRC)gsargs.h:$(GLGEN)arch.h
+$(GLSRC)gsargs.h:$(GLSRC)stream.h
 $(GLSRC)gsexit.h:$(GLSRC)std.h
 $(GLSRC)gsexit.h:$(GLSRC)stdpre.h
 $(GLSRC)gsexit.h:$(GLGEN)arch.h
diff --git a/base/mkromfs.c b/base/mkromfs.c
index 2b9d4363d..f0f98543d 100644
--- a/base/mkromfs.c
+++ b/base/mkromfs.c
@@ -147,6 +147,19 @@ static inline int isbigendian(void)
     return u.c[0] != 1;
 }
 
+/* The gp_local_arg_encoding_get_codepoint() function presnet in (some) OS-specific
+ * files (gp_unix.c, gp_win32.c) now uses streams, so it won't compile unless there
+ * is an implementation of spgetcc(). We don't actually use the function for mkromfs
+ * so we just need to make sure it compiles.
+ */
+int spgetcc(stream *s, bool b)
+{
+    (void) s; /* avoid 'unused' warning from compilers */
+    (void) b;
+
+    return 0;
+}
+
 /* mkromfs doesn't use gp_stat, but it does link gp_misc.c which includes
    call to gp_stat_impl(). Rather than major build upheaval for something not
    used, just define a dummy here for Windows.
diff --git a/base/unix-aux.mak b/base/unix-aux.mak
index bf99697b8..38ddd109c 100644
--- a/base/unix-aux.mak
+++ b/base/unix-aux.mak
@@ -37,7 +37,7 @@ $(GLGEN)unix_.dev: $(unix__) $(GLD)nosync.dev $(GLD)smd5.dev $(UNIX_AUX_MAK) $(M
 	$(ADDMOD) $(GLGEN)unix_ -include $(GLD)smd5
 
 $(GLOBJ)gp_unix.$(OBJ): $(GLSRC)gp_unix.c $(AK)\
- $(pipe__h) $(string__h) $(time__h) $(gx_h) $(gsexit_h) $(gp_h) $(UNIX_AUX_MAK) $(MAKEDIRS)
+ $(pipe__h) $(string__h) $(time__h) $(gx_h) $(gsexit_h) $(gp_h) $(stream_h) $(UNIX_AUX_MAK) $(MAKEDIRS)
 	$(GLCC) $(FONTCONFIG_CFLAGS) $(GLO_)gp_unix.$(OBJ) $(C_) $(GLSRC)gp_unix.c
 
 $(AUX)gp_unix.$(OBJ): $(GLSRC)gp_unix.c $(AK)\
diff --git a/base/winplat.mak b/base/winplat.mak
index 0fc57698f..0be1e4801 100644
--- a/base/winplat.mak
+++ b/base/winplat.mak
@@ -50,12 +50,12 @@ $(AUX)gp_ntfs.$(OBJ): $(GLSRC)gp_ntfs.c $(AK)\
 
 $(GLOBJ)gp_win32.$(OBJ): $(GLSRC)gp_win32.c $(AK)\
  $(dos__h) $(malloc__h) $(stdio__h) $(string__h) $(windows__h)\
- $(gp_h) $(gsmemory_h) $(gstypes_h) $(WINPLAT_MAK)
+ $(gp_h) $(gsmemory_h) $(gstypes_h) $(stream_h) $(WINPLAT_MAK)
 	$(GLCCWIN) $(GLO_)gp_win32.$(OBJ) $(C_) $(GLSRC)gp_win32.c
 
 $(AUX)gp_win32.$(OBJ): $(GLSRC)gp_win32.c $(AK)\
  $(dos__h) $(malloc__h) $(stdio__h) $(string__h) $(windows__h)\
- $(gp_h) $(gsmemory_h) $(gstypes_h) $(WINPLAT_MAK)
+ $(gp_h) $(gsmemory_h) $(gstypes_h) $(stream_h) $(WINPLAT_MAK)
 	$(GLCCAUX) $(AUXO_)gp_win32.$(OBJ) $(C_) $(GLSRC)gp_win32.c
 
 # Define the Windows thread / synchronization module.
diff --git a/psi/files.h b/psi/files.h
index 68d3395c6..7ad161b81 100644
--- a/psi/files.h
+++ b/psi/files.h
@@ -103,7 +103,7 @@ typedef struct gs_file_path_s *gs_file_path_ptr;
 
 /* Procedures exported by zfile.c. */
         /* for imainarg.c */
-gp_file *lib_fopen(const gs_file_path_ptr pfpath, const gs_memory_t *mem, const char *);
+stream *lib_sopen(const gs_file_path_ptr pfpath, const gs_memory_t *mem, const char *);
 
         /* for imain.c */
 int
diff --git a/psi/imainarg.c b/psi/imainarg.c
index aaf41b6bf..b3269e2e7 100644
--- a/psi/imainarg.c
+++ b/psi/imainarg.c
@@ -114,11 +114,11 @@ static void print_help_trailer(const gs_main_instance *);
 /* ------ Main program ------ */
 
 /* Process the command line with a given instance. */
-static gp_file *
-gs_main_arg_fopen(const char *fname, void *vminst)
+static stream *
+gs_main_arg_sopen(const char *fname, void *vminst)
 {
     gs_main_set_lib_paths((gs_main_instance *) vminst);
-    return lib_fopen(&((gs_main_instance *)vminst)->lib_path,
+    return lib_sopen(&((gs_main_instance *)vminst)->lib_path,
                      ((gs_main_instance *)vminst)->heap, fname);
 }
 static void
@@ -140,7 +140,7 @@ gs_main_init_with_args01(gs_main_instance * minst, int argc, char *argv[])
 
     /* Now we actually process them */
     code = arg_init(&args, (const char **)argv, argc,
-                    gs_main_arg_fopen, (void *)minst,
+                    gs_main_arg_sopen, (void *)minst,
                     minst->get_codepoint,
                     minst->heap);
     if (code < 0)
diff --git a/psi/imainarg.h b/psi/imainarg.h
index 7e330b901..fe94a77e9 100644
--- a/psi/imainarg.h
+++ b/psi/imainarg.h
@@ -49,7 +49,7 @@ int gs_main_run_start(gs_main_instance * minst);
  * Allocation should be done with the standard gs allocation functions, as
  * the returned string may be freed using the same.
  */
-typedef int (gs_arg_get_codepoint)(gp_file *, const char **);
+typedef int (gs_arg_get_codepoint)(stream *, const char **);
 
 void gs_main_inst_arg_decode(gs_main_instance * minst,
                              gs_arg_get_codepoint *get_codepoint);
diff --git a/psi/iminst.h b/psi/iminst.h
index 4e71056d2..b249def51 100644
--- a/psi/iminst.h
+++ b/psi/iminst.h
@@ -77,7 +77,7 @@ struct gs_main_instance_s {
     long base_time[2];		/* starting usertime */
     void *readline_data;	/* data for gp_readline */
     ref error_object;		/* Use by gsapi_*() */
-    int (*get_codepoint)(gp_file *file, const char **astr);
+    int (*get_codepoint)(stream *s, const char **astr);
                                 /* Get next 'unicode' codepoint */
     display_callback *display;	/* callback structure for display device */
     /* The following are updated dynamically. */
diff --git a/psi/psapi.c b/psi/psapi.c
index 9321f34a9..454aef1e9 100644
--- a/psi/psapi.c
+++ b/psi/psapi.c
@@ -165,7 +165,7 @@ psapi_delete_instance(gs_lib_ctx_t *ctx)
         --gsapi_instance_counter;
 }
 
-static int utf16le_get_codepoint(gp_file *file, const char **astr)
+static int utf16le_get_codepoint(stream *s, const char **astr)
 {
     int c;
     int rune;
@@ -181,11 +181,11 @@ static int utf16le_get_codepoint(gp_file *file, const char **astr)
      * be wrong. */
 
     do {
-        if (file) {
-            rune = gp_fgetc(file);
+        if (s) {
+            rune = spgetc(s);
             if (rune == EOF)
                 return EOF;
-            c = gp_fgetc(file);
+            c = spgetc(s);
             if (c == EOF)
                 return EOF;
             rune += c<<8;
@@ -207,11 +207,11 @@ static int utf16le_get_codepoint(gp_file *file, const char **astr)
 lead: /* We've just read a leading surrogate */
         rune -= 0xD800;
         rune <<= 10;
-        if (file) {
-            trail = gp_fgetc(file);
+        if (s) {
+            trail = spgetc(s);
             if (trail == EOF)
                 return EOF;
-            c = gp_fgetc(file);
+            c = spgetc(s);
             if (c == EOF)
                 return EOF;
             trail += c<<8;
diff --git a/psi/zfile.c b/psi/zfile.c
index e4a2a3e17..b8a17c8a7 100644
--- ghostscript-9.50/psi/zfile.c.orig	2019-10-15 11:05:37.000000000 +0200
+++ ghostscript-9.50/psi/zfile.c	2024-06-03 14:46:30.167002538 +0200
@@ -1274,8 +1274,8 @@
 }
 
 /* The startup code calls this to open @-files. */
-gp_file *
-lib_fopen(const gs_file_path_ptr pfpath, const gs_memory_t *mem, const char *fname)
+stream *
+lib_sopen(const gs_file_path_ptr pfpath, const gs_memory_t *mem, const char *fname)
 {
     /* We need a buffer to hold the expanded file name. */
     char filename_found[DEFAULT_BUFFER_SIZE];
@@ -1290,8 +1290,8 @@
 
     if (code < 0)
         return NULL;
-    file = ((stream *)(obj.value.pfile))->file;
-    return file;
+
+    return((stream *)(obj.value.pfile));
 }
 
 /* Open a file stream that reads a string. */
-- 
2.44.0.windows.1

